prov/util/src/util_attr.c

```diff
@@ -56,11 +57,167 @@ static int fi_valid_addr_format(uint32_t prov_format, uint32_t user_format)
 	}
 }
 
+/*
+ * Accepts a string which has tokens delimited by underscores and returns a token
+ * array, tok[0..max_token - 1]. Callers are responsible for freeing tok[0] which
+ * will point to a fudged copy of the string. Refer strtok_r manpage for details.
+ */
+int utilx_parse_name(const char *name, int max_tok, char **tok, int get_prefix)
+{
+	char *name_dup, *name_ref, *save_ptr;
+	char *delim = "_";
+	int i, parsed_len = 0;
+
+	name_dup = strdup(name);
+	if (!name_dup)
+		return -FI_ENOMEM;
+
+	name_ref = name_dup;
+	for (i = 0; i < max_tok; name_ref = NULL, i++) {
+		tok[i] = strtok_r(name_ref, delim, &save_ptr);
+		if (!tok[i])
+			break;
+	}
+	if (i < max_tok)
+		goto err;
+
+	if (!get_prefix) {
+		/* Handle the case where there is a delim in token itself */
+		for (i = 0; i < max_tok; i++)
+			parsed_len += strlen(tok[i]) + 1;
+		parsed_len--;
+
+		if (strlen(name) > parsed_len)
+			tok[max_tok - 1][strlen(tok[max_tok - 1])] = *delim;
+	}
+
+	return 0;
+err:
+	free(name_dup);
+	return -FI_EOTHER;
+}
+
+int utilx_tr_layer_domain_attr(struct fi_domain_attr *layer_attr,
+		struct fi_domain_attr *base_attr)
+{
+	char *tok[2];
+	int ret;
+
+	ret = utilx_parse_name(layer_attr->name, 2, tok, 0);
+	if (ret)
+		return ret;
+	if (!(base_attr->name = strdup(tok[1])))
+		ret = -FI_ENOMEM;
```

Comment:
We do know domain_attr->name prefix for the layered provider. I think we could simplify this to use strdup where the string starting position is offset by strlen(prefix) + 1.